src/common/barcode_index/barcode_index.hpp

```diff
@@ -0,0 +1,823 @@
+ * Read cloud is represented by its barcode
+ * The edge contains the cloud if there is a read barcoded by cloud's barcode which is aligned to the edge.
+ * Info example: FrameBarcodeInfo
+ */
+template <class Graph, class EdgeEntryT>
+class BarcodeIndex: public AbstractBarcodeIndex<Graph> {
+friend class BarcodeIndexInfoExtractor<Graph, EdgeEntryT>;
+
+public:
+    typedef typename Graph::EdgeId EdgeId;
+    typedef typename Graph::VertexId VertexId;
+    typedef typename omnigraph::IterationHelper <Graph, EdgeId> edge_it_helper;
+    typedef std::unordered_map <EdgeId, EdgeEntryT> barcode_map_t;
+
+    BarcodeIndex (const Graph &g) :
+            AbstractBarcodeIndex<Graph>(g),
+            edge_to_entry_(),
+            number_of_barcodes_(0)
+    {}
+
+    BarcodeIndex (const BarcodeIndex& other) = default;
+
+    virtual ~BarcodeIndex() {}
+
+    size_t size() const {
+        return edge_to_entry_.size();
+    }
+
+    bool empty() const {
+        return size() == 0;
+    }
+
+    typename barcode_map_t::iterator begin() noexcept {
+        return edge_to_entry_.begin();
+    }
+    typename barcode_map_t::iterator end() noexcept {
+        return edge_to_entry_.end();
+    }
+    typename barcode_map_t::iterator begin() const noexcept {
+        return edge_to_entry_.begin();
+    }
+    typename barcode_map_t::iterator end() const noexcept {
+        return edge_to_entry_.end();
+    }
+    typename barcode_map_t::const_iterator cbegin() const noexcept {
+        return edge_to_entry_.cbegin();
+    }
+    typename barcode_map_t::const_iterator cend() const noexcept {
+        return edge_to_entry_.cend();
+    }
+
+    size_t GetBarcodeNumber(const EdgeId &edge) const override {
+        return GetEntry(edge).Size();
+    }
+
+    bool IsEmpty() override {
+        return size() == 0;
+    }
+
+    //Delete low abundant barcodes from every edge
+    void Filter(size_t trimming_threshold, size_t gap_threshold) override {
+        for (auto entry = edge_to_entry_.begin(); entry != edge_to_entry_.end(); ++entry) {
+            entry->second.Filter(trimming_threshold, gap_threshold);
+        }
+    }
+
+    void ReadEntry (std::ifstream& fin, const EdgeId& edge) override {
+        DEBUG("Reading entry")
+        DEBUG("Edge: " << edge.int_id());
+        DEBUG("Length: " << g_.length(edge));
+        edge_to_entry_[edge].Deserialize(fin);
+    }
+    void WriteEntry (std::ofstream& fout, const EdgeId& edge) override {
+        fout << g_.int_id(edge) << std::endl;
+        GetEntry(edge).Serialize(fout);
+    }
+
+    virtual void BinRead(std::istream &str) {
+        using io::binary::BinRead;
+
+        edge_to_entry_.clear();
+        size_t size;
+        BinRead(str, size);
+        for (size_t i = 0; i < size; ++i) {
+            EdgeId edge_id = BinRead<uint64_t>(str);
+            auto entry = BinRead<EdgeEntryT>(str);
+            edge_to_entry_.insert({std::move(edge_id), std::move(entry)});
```

Comment:
This does not make much sense as:
 - edgeid is a trivial type, so move is equivalent to copy
 - You're creating a pair

You'd better do:
```c++
edge_to_entry_.emplace(edge_id, std::move(entry));
```

or simply:
```c++
edge_to_entry.emplace(edge_id, BinRead<EdgeEntryT>(str));
```
(do not do two BinReads though as order of evaluation of arguments is arbitrary)